### ISSUE TITLE
Breaking change (maybe?) in minor version update `2.3.x` to `2.4.x`

### ISSUE THREAD

**Issue by jjoelson** [simolus3/moor#2252]
Hi all! Our team encountered what looks like a breaking change between drift `2.3.x` and `2.4.x`. The reason that I put "maybe" in the title is because the definition of "breaking change" is a bit murky when talking about code generation, and I'm not sure exactly what backward compatibility guarantees this project intends.

The issue we found is that code generated at build time with `drift_dev: 2.4.x` has an error when running an app with `drift: 2.3.0` pertaining to the type converter for an integer enum column:
```
Error: A value of type 'EnumIndexConverter<MyEnum>' can't be assigned to a variable of type 'JsonTypeConverter2<MyEnum, int, int>'
```

If this type of backwards compatibility is not intended, then we will simply lock the drift dependencies to a specific version to prevent any problems.

Thanks!

**Comment by simolus3** [comment]
Yes, the intention is that the "internal API" used between `drift` and `drift_dev`-generated code can change in minor (but not in patch) versions. Otherwise, it would be hard to introduce features that affect both the main package and the generator since the generator would have to assume it's running against the oldest possible `drift` package.

> The issue we found is that code generated at build time with `drift_dev: 2.4.x` has an error when running an app with `drift: 2.3.0`

I've double-checked that `drift_dev` versions 2.4.x have `drift: >=2.4.0 < 2.5.0` as a dependency, so it should be impossible for `pub get` to pick versions of `drift` and `drift_dev` with different non-patch versions. Do you have a `dependency_override` for either package? I'm a bit surprised to hear that you ran into this, I thought the tight constraints in `drift_dev` would prevent this.

**Comment by jjoelson** [comment]
Thanks for the quick response @simolus3!

It's possible we're handling our lock files incorrectly. We have a mono repo that includes an app package and a library package that contains the Drift codegen stuff:
- `my_app`
- `my_database`

In accordance with the [Flutter guidance for `pub spec.lock`](https://dart.dev/guides/libraries/private-files#pubspeclock), the lock file is committed to the repository for `my_app`, but not for `my_database`. 

So what happened is `my_app`'s `pubspec.lock` had drift `2.3.0` committed to the repo, but `my_database`'s `pubspec.lock` had drift and drift_dev at either `2.3.x` or `2.4.x` depending on when each developer first ran `pub get`; if they did it prior to the release of drift `2.4` then they got `2.3` and everything worked properly, but if they did it _after_ the release of drift `2.4` then they got `2.4` and the generated code was no longer compatible with `my_app`s drift version. And CI builds, which clone the repo fresh each time, stopped working when `2.4` was released.

Specifying a single version in `my_database` (i.e. `drift: 2.3.0` instead of `drift: ^2.3.0`) is the only way I can think of to fully avoid this situation so long as the code-gen is happening in a different package from the code that will be using it. 

Have you encountered this scenario before?

**Comment by simolus3** [comment]
Thanks for the follow-up. I think I want to keep the current behavior of allowed internal breakages in minor drift versions. This also seems to be what other code-generation packages do, e.g. [built_value here](https://github.com/google/built_value.dart/blob/3e26f69643a98f1746d6f1ee7693c42e072deb18/built_value_generator/pubspec.yaml#L16).

I understand the problem, but honestly I don't know if that's something that can realistically be solved in drift. It's a perfectly reasonable CI setup, but I also think the contract we have in drift is reasonable. Even without multiple packages in a repository, non-breaking changes in `drift_dev` could break a CI. For instance, users could be using [build_verify](https://pub.dev/packages/build_verify) to ensure developers keep generated sources up-to-date. A minor change in `drift_dev` (say to optimize some generated code) would break this CI check. So it seems like it seems like there are reasonable CI setups that we will break with changes we don't consider breaking.

It would be great if there was some tool that manages dependencies for monorepos so that all packages inside get the same version for dependencies, but I'm not aware of such a solution.

I'll keep this issue open to see if anyone else has suggestions that could improve this situation, but I don't have a good plan that would address this issue unfortunately.

**Comment by jjoelson** [comment]
Thanks for explaining the reasoning behind the API contract. I agree that allowing "internal breakages" in minor version updates is the right move because it works best for the common case of application packages that depend on `drift` directly; the ability to release optimizations for the generated code in minor updates is valuable!

But if I'm understanding the situation correctly, I think it's not just mono repos that could be affected; any library package that uses Drift internally to manage a database needs to be careful about this. For example, if `my_package` has `drift: ^2.3.0` in its `pubspec.yaml` and contains code generated by `drift_dev` version `2.3.0`, then `my_package` is in violation of semantic versioning: it claims to be compatible with any drift version up to the next major version, but would actually break if used with `2.4.0`. 

Does `drift` guarantee no "internal API" breaking changes in patch versions? If so, I think the proper thing for `my_package` to have in its `pubspec.yaml` in this scenario is `drift: ">=2.3.0 <2.4.0"`.

It might be worth putting something in the docs about this for people who plan to use Drift in a library package. If you agree, I could try writing something up and submitting a PR. Let me know if that would be helpful!

**Comment by simolus3** [comment]
> It might be worth putting something in the docs about this for people who plan to use Drift in a library package.

I agree, this should be mentioned in the documentation. A PR would be most welcome, thanks for your help!
